### PR TITLE
Zoltan2: Change resetParameters and all constructors have comm #1014 #1004

### PR DESCRIPTION
**Pull request by MicheldeMessieres**
Changed the way resetParameters() is setup as discussed.
This removed the old change and implemented a new one.
This is not expected to change behavior.

Turns on rcb_C unit test always.

Changed default Environment constructor to take a comm.
Changed all code dependent on default Environment().
Generally this meant:
  CHanging a lot of main() functions to simply pass comm.
  Changing a few error messages to be handled manually instead of with Env.

Note that in Zoltan2_TaskMapping.h I changed an API to take an
Environment as I believe the problemComm should not be used in this place.
This is called from taskMappingTest3.cpp.
However I am not sure if this is desirable or even ok to change that API.

### CODE REVIEW

File: packages/zoltan2/src/algorithms/partition/Zoltan2_TaskMapping.hpp [hunk 1]
```diff
@@ -2849,6 +2849,7 @@ class CoordinateTaskMapper:public PartitionMapping<Adapter>{
  */
 template <typename part_t, typename pcoord_t, typename tcoord_t>
 void coordinateTaskMapperInterface(
+    const Teuchos::RCP <const Environment> envConst_,
```
Comment (kddevin) [inline packages/zoltan2/src/algorithms/partition/Zoltan2_TaskMapping.hpp]: Changing the interface will impact users.  I suggest deleting this change.

File: packages/zoltan2/src/algorithms/partition/Zoltan2_TaskMapping.hpp [hunk 2]
```diff
@@ -2868,9 +2869,6 @@ void coordinateTaskMapperInterface(
     bool divide_to_prime_first = false
 )
 {
-
-  const Environment *envConst_ = new Environment();
```
Comment (kddevin) [inline packages/zoltan2/src/algorithms/partition/Zoltan2_TaskMapping.hpp]: I would keep this line, and use problemComm as the communicator.

File: packages/zoltan2/src/algorithms/partition/Zoltan2_TaskMapping.hpp [hunk 3]
```diff
@@ -2920,8 +2918,6 @@ void coordinateTaskMapperInterface(
     proc_to_task_adj[i] = proc_to_task_adj_[i];
   }
   delete ctm;
-  delete envConst_;
```
Comment (kddevin) [inline packages/zoltan2/src/algorithms/partition/Zoltan2_TaskMapping.hpp]: If keep the construction, keep this destruction as well.

File: packages/zoltan2/test/partition/taskMappingTest3.cpp [hunk 4]
```diff
@@ -400,7 +400,10 @@ int main(int argc, char *argv[]){
         //hopper[2] = 24;
         part_t *machineDimensions = NULL;
         //machineDimensions = hopper;
+        Teuchos::RCP <Zoltan2::Environment> env
+          = rcp(new Zoltan2::Environment(comm));
```
Comment (kddevin) [inline packages/zoltan2/test/partition/taskMappingTest3.cpp]: In general, we don't want to force users to create the environment.  I propose returning to original coordinateTaskMapperInterface, and using the problemComm as the communicator to create the needed environment.

**Comment by kddevin** [comment]
Regarding the very old comment regarding the communicator in Zoltan2_Environment.hpp:
"   *   Note that the communicator is for the application, not the problem."
The way the environment is used, I cannot see why we would want a different communicator in the environment from that in the problem.  The communicator is used for global error checking within the problem and algorithms; thus, it needs to contain the same processors as the problem.  Is there another use case?  If not, the comments should probably be removed to prevent further confusion.

**Comment by MicheldeMessieres** [comment]
@kddevin New changes restore the original TaskMapping setup except for the one change to make Environment built with problemComm. I also assumed it would be appropriate to go ahead and remove the comment in Zoltan2_Environment.hpp regarding differences between application and problem comm.

**Comment by kddevin** [comment]
Looks good.  Thanks, @MicheldeMessieres 
We are now xSDK compliant. :)

**Pull request merged**